### DISCUSSION TITLE
MapStorageProvider SPI to replace custom user storage spis ?

### DISCUSSION THREAD

**Discussion by alplunge** [keycloak/keycloak#14142] [Archive]
With the release of keycloak 19, there were interesting announcements made:

> **Creating custom storage providers**
> The API for creating a custom storage provider has not been fully stabilized yet, though it is available as a tech preview. See the MapStorageProvider SPI and its Javadoc for details. The availability of the new API is a priority for the next Keycloak version.

Also

> User storage SPI will become deprecated. It will be supported for several more versions, but will be eventually replaced by the Map Storage SPI, which will offer the ability to create custom storages for any recognized area, such as users, roles, clients, or groups.

I am currently using my very own custom user storage spi, build based on this [user storage guide](https://www.keycloak.org/docs/latest/server_development/#_user-storage-spi) which also states

> This functionality depends on APIs bundled in the keycloak-model-legacy module. It will soon be replaced with the new map storage API which provides a uniform way to access both local and external information about users and other entities, and the old APIs will be removed eventually.

As this new `Map Storage SPI` still just in the **tech preview** mode, could we simple mortal developers expect migration documentation on how to integrate with `Map Storage SPI` API, or updated version of [user storage guide](https://www.keycloak.org/docs/latest/server_development/#_user-storage-spi) ?

Thank you for your responses 🙏

**Comment by alplunge** [comment]
it seems that I am the only one having these particular questions... 🙄

**Reply by ahus1** [reply]
Hi @alplunge - eventually there will be some docs. For now, there are currently three implementations for for a map user storage: the concurrent hash map implementation, the JPA and the Hot Rot/Infinispan solution. 

Each of those can replace the storage of user information in a Keycloak installation across all realms. If this is what you need, I might be able to give some guidance and explain the existing implementations if you want to explore them. This won't replace a guide, still it could be a start if you want to give it a try.

As part of the development journey of the map storage, we also plan have the storage configurable per realm, so you can have one realm where an entity like users is backed by a JPA store, and anther realm where it is backed by a JPA store in another database, in Hot Rot or a different implementation. 

Once the tree store functionality arrives as part of the map store, different attributes of an entity can be stored in different stores. This comes in handy when handling for example users in LDAP where only some information is stored in LDAP, and other attributes that are required for Keycloak are stored in JPA. 

Depending on the needs of your user storage, it might necessary to either wait for the functionality to arrive, or you might be able to start exploring now. Please let me know your needs so that we can find out if the map store is ready yet.

Finally the usual disclaimer: all official road map announcements will be done in the blog, don't take this as a road map. See https://www.keycloak.org/2022/07/storage-map for a roadmap.

**Reply by alplunge** [reply]
@ahus1 thank you for your detailed reply. The usage of custom user storage spi is this:
when requesting for an access token (on the specific realm, but this logic should be consistent with any realm), keycloak consults user storage spi to fetch user information (actual user and its roles) from the external storage by passing needed info about the user from access token request. So User storage spi makes that connection on behalf of keycloak and queries that user info from external DB, once that info is fetched it maps to appropriate keycloak's user and roles representations so that keycloak can make sense out of the users data in order to generate access token..

**Reply by ahus1** [reply]
@alplunge - thanks for describing the user storage SPI. My question might have been a bit vague, so to rephrase it: I'd love to hear about more what you've built with the user storage SPI to understand if the map storage is ready in its current state to support what you're already doing or might want to do in the future:

* Would you need a realm specific configuration?
* Is all information about the user available from the external database?
* How do you handle credentials?

Again, the way you handle it today might be different from how this might be handed in the map storage provider, so it is good to understand the business need and not only look at the technical implementation.

**Reply by alplunge** [reply]
@ahus1 so that description is exactly what custom user storage spi does at the moment. Which was built based [user storage guide](https://www.keycloak.org/docs/latest/server_development/#_user-storage-spi).

> Would you need a realm specific configuration?

Each realm would need to use exact same logic as described in my previous comments. The only difference I would say, that each realm has its own unique client, which would be used when making `.../protocol/openid-connect/token` but I guess this is irrelevant atm. 

> Is all information about the user available from the external database?

Indeed. Keycloak in this scenario relies heavily on SPI for user information and only keeps that user info for some time in the local cache which is done by default. When it comes to additional attributes of a user, which cannot be mapped to an keycloak representation, then `CachedUserModel` is used.

> How do you handle credentials?

credentials of a fetched user from external DB are not really relevant atm. as those users are kinda specific ones, and password is not that important at the moment. In order for client to be authorised to make a `.../protocol/openid-connect/token` on the specific realm, JWKS mechanism is in use. 


> Again, the way you handle it today might be different from how this might be handed in the map storage provider, so it is good to understand the business need and not only look at the technical implementation.

I think that is completely understandable as long as there is some guide to follow through :)

**Reply by alplunge** [reply]
@ahus1 is my reply sufficient enought to carry on disucssion about how to replace user storage spi with a new map storage spi so that it can be used for User Federation?

**Reply by ahus1** [reply]
@alplunge - with the answers above, using the map storage provider for users might or might not work in your case. I can't fully assess all the information from above. 

Please note: the classic User Storage Providers can be activated per realm. The new map storage doesn't support that yet: If you go that route today, your provider need to handle the users of all realms (including the master realm). This will change before the map storage graduates from the preview state.

PS: I think you misspelled my GitHub username above...

**Reply by alplunge** [reply]
@ahus1 Thank you for clarifying. will postpone user storage migration using new map storage, till map storage graduates from the preview state.

P.S.: sorry for misspelled username! ✋

**Comment by dishi-github** [comment]
The following dependencies although change the methods and their signature. The change in the order of methods parameters is mentioned in the updated doc and is not a big deal however, my concerns are the following:

1. I had this class CustomUser to map the external users with Keycloak. It was extending another class, whose methods have changed and I am supposed to implement this method called ` public SubjectCredentialManager credentialManager()`

Can someone guide me as to what goes inside this method? I checked other implementations of this method in the legacy jar, but I am clueless. 

2. The methods have been renamed of this interface `UserStorageProvider`. Can someone help me with the proper parameter names?  Any mapping between old method names and new ones?

3. Last concern is that, do I need all of these dependencies. What is the purpose of each?

Dependencies :

     
       <dependency>
            <groupId>org.keycloak</groupId>
            <artifactId>keycloak-model-legacy</artifactId>
            <version>${keycloak.version}</version>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.keycloak</groupId>
            <artifactId>keycloak-model-legacy-services</artifactId>
            <version>${keycloak.version}</version>
            <scope>provided</scope>
        </dependency>

        <dependency>
            <groupId>org.keycloak</groupId>
            <artifactId>keycloak-model-legacy-private</artifactId>
            <version>${keycloak.version}</version>
            <scope>provided</scope>
        </dependency>
```

**I am trying to migrate from Keycloak 18.0.2 to 20.0.0.**

**Reply by ahus1** [reply]
Hi @dishi-github - thanks for reaching out. You're mentioning that you're existing class extends a class, but didn't name it here. Assuming that you are extending `UserAdapter`, there is already a default implementation that might suite your needs. If you're not using the new map store, you'll probably need those dependencies to extend the classes you choose to extend.

The `SubjectCredentialManager` is all about credential: storing and validating them. Credentials range from username and password to token generators. So you can choose to delegate some or all of this functionality to the external system that stores your users.
